### DISCUSSION TITLE
TypeScript Error with UseFieldArrayReturn and Generic Type Parameters

### DISCUSSION THREAD

**Discussion by bitboxx** [react-hook-form/react-hook-form#11746] [Q&A]
Hello,

I've encountered a TypeScript error when trying to use `UseFieldArrayReturn` with a generic type parameter. The issue does not occur when using a non-generic type. Here's a brief overview:

```typescript
import { UseFieldArrayReturn } from 'react-hook-form'

type ArrFieldType = {
  id: string
  name: string
}

type ListType = { list: ArrFieldType[] }

// Works as expected
type WithoutGenericsType = UseFieldArrayReturn<ListType, 'list', 'fieldKey'>

// TypeScript error with generic type
type WithGenericsType<T extends ListType> = UseFieldArrayReturn<
  T, 
  'list', // Error: Type '"list"' does not satisfy the constraint 'ArrayPath<T>'.
  'fieldKey'
>
```

The error specifically states: Type '"list"' does not satisfy the constraint 'ArrayPath<T>'.

I followed the TypeScript integration guide from the official documentation (https://react-hook-form.com/ts#UseFieldArrayReturn) and expected the generic version to work similarly to the non-generic version without errors.

Could you please provide insight or a solution for this issue? Is there an alternative approach to using `UseFieldArrayReturn` with generic type parameters, or is this an oversight in the type definitions?

Here's a reproduction of the issue: https://stackblitz.com/edit/react-hook-form-8heqxa?file=src%2FApp.tsx&terminal=dev

**Comment by sujinleeme** [comment]
I guess you miss `'list' & FieldArrayPath<T>`.

[[playground]](BXAZyQDFgkAbAEwCCUKCgCeAGjgBVdl16DhY5DGZQAdpLn8hI0QAUUMABZwAvnGy4CRdDAC0RiBADWdzNBD4qVGKLBI4ABlgVngAXkY4HhCYAC5GKjg4YD540KhgNQBzRLg1FBAkNJgM7KpTAG0AXTNvSl9-OAB1YGMAcSQ1JAy0VgAVPyQAHj64JAAPGE6+ViCYgD44CJlObm1FUWVVNSHcvvFc-GjQ-DgAMjgtBV0DYxH5qgfqesG4Dq6elaudJSQVdQAjEtmq0jO9usBegN-EMGLkklkcMwwAA5ApFAiIiDIuwA-AHShJJLHOIJQlEikpYqlLIEikU-KFamZHLkpKVKrlJ5UAD0PJBxmSaimUHYGGAEDUPle4M+sjW11+-zUACZgS12p0IVDBrD4XAsci0UzMUiwHYVfjyuc5qFuc8+XBoQE0EYkGhnCyqGhJaE4IVjBBUm8tXLVvIfps-ts1RE2OH1roABRMRkYo4xU6mACUVCAA)


```ts
import React from 'react';

import { useFieldArray, UseFieldArrayReturn, FieldArrayPath } from 'react-hook-form'

type List = { list: {
  id: string
  name: string
}[] }

type WithGenericsType<T extends List> = UseFieldArrayReturn<
  T,
  'list' & FieldArrayPath<T>
>

type GenericUseFieldArrayReturn1 = WithGenericsType<{
    groupName: 'group-1',
    list: {
        id: string,
        name: string
    }[]
}>

// With intersection
type GenericUseFieldArrayReturn2 = WithGenericsType<{
    groupName: 'group-2',
} & List>


// Type checking
const method: GenericUseFieldArrayReturn2 = useFieldArray({ name: 'list' })
```

**Comment by bitboxx** [comment]
Thanks @sujinleeme. Your idea to use `'list' & FieldArrayPath<T>` did help resolve the TS error. However, it seems like this approach might inadvertently widen the types, which could lead to less strict type checking than intended.

I had a discussion with my colleague about your solution and experimented a bit. I found an alternative that might be more precise:

```typescript
type WithGenericsType<T extends ListType> = UseFieldArrayReturn<
  T,
  T['list'], // Using the type of 'list' directly from T
  'fieldKey'
>;
```

This keeps the type strictly tied to `T`'s structure, not widening the type check to any field in T (& `list`).

Thought this might be interesting for you as well.

**Reply by sujinleeme** [reply]
`T['list']` makes sense! Thanks for sharing this 👍
